lib/ddtrace/contrib/racecar/patcher.rb

```diff
@@ -0,0 +1,78 @@
+require 'ddtrace/ext/app_types'
+
+module Datadog
+  module Contrib
+    module Racecar
+      # Provides instrumentation for `racecar` through ActiveSupport instrumentation signals
+      module Patcher
+        include Base
+        NAME = 'racecar.consumer'.freeze
+        register_as :racecar
+        option :tracer, default: Datadog.tracer
+        option :service_name, default: 'racecar'
+
+        class << self
+          def patch
+            return patched? if patched? || !compatible?
+
+            ::ActiveSupport::Notifications.subscribe('start_process_batch.racecar', &method(:start))
+            ::ActiveSupport::Notifications.subscribe('start_process_message.racecar', &method(:start))
+            ::ActiveSupport::Notifications.subscribe('process_batch.racecar', &method(:finish))
+            ::ActiveSupport::Notifications.subscribe('process_message.racecar', &method(:finish))
+
+            configuration[:tracer].set_service_info(
+              configuration[:service_name],
+              'racecar',
+              Ext::AppTypes::WORKER
+            )
+
+            @patched = true
+          end
+
+          def patched?
+            return @patched if defined?(@patched)
+            @patched = false
+          end
+
+          private
+
+          def configuration
+            Datadog.configuration[:racecar]
+          end
+
+          def compatible?
+            defined?(::Racecar) && defined?(::ActiveSupport::Notifications)
+          end
+
+          def start(*_, payload)
+            ensure_clean_context!
+
+            span = configuration[:tracer].trace(NAME)
+            span.service = configuration[:service_name]
+            span.resource = payload[:consumer_class]
+            span.set_tag('kafka.topic', payload[:topic])
+            span.set_tag('kafka.consumer', payload[:consumer_class])
+            span.set_tag('kafka.partition', payload[:partition])
+            span.set_tag('kafka.offset', payload[:offset]) if payload.key?(:offset)
+            span.set_tag('kafka.first_offset', payload[:first_offset]) if payload.key?(:first_offset)
```

Comment:
This mixes up batch processing and individual message processing. For the former, you'd be interested in

1. The start and end offsets
2. The batch size

For the latter you're just interested in the offset of that specific message.

Reply:
I'm conditionally setting the fields `kafka.offset` and `kafka.first_offset`.
* For the `process_message.racecar` event you provide a payload with `consumer_class`, `topic`,  `partition` and `offset` so these are set in the span accordingly.
* For the `process_batch.racecar`, event you provide a payload with `consumer_class`, `topic`,  `partition` and `first_offset` which are the fields set in batch consumer spans.
Let me know if that makes sense!

Reply:
Sure, I was just confused by both message and batch processing was mixed together. For batch processing, the number of messages is obviously important.

Reply:
So do you plan to update the instrumentation payload with that field?

Reply:
Ah, thought I already had that! I'll add it now.

Reply:
https://github.com/zendesk/racecar/pull/47

You can add the field conditionally, I'll merge this ASAP.

Reply:
Cool, thank you!